{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Hi-C class script\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from scipy.stats import zscore\n",
    "import argparse\n",
    "import itertools\n",
    "\n",
    "data_path = \"/Users/ninaxiong/projects/HST508_Fall2022_final_project/data/ENCODE_bulk_rna_seq.csv\"\n",
    "gene_loc_path = \"/Users/ninaxiong/projects/HST508_Fall2022_final_project/data/gene_locations.tsv\"\n",
    "tad_path = \"/Users/ninaxiong/projects/HST508_Fall2022_final_project/data/TAD_strong_boundary_start_end.csv\"\n",
    "chrlen_path = \"/Users/ninaxiong/projects/HST508_Fall2022_final_project/data/chr_lengths\"\n",
    "\n",
    "def extract_data(data_path,gene_loc_path,tad_path,excl_chrom=['chrM','chrX','chrY']):\n",
    "    \"\"\"\n",
    "    excl_chrom: list of strings corresponding to chromosomes to be excluded\n",
    "    data_path: path to single cell tpm data\n",
    "    gene_loc_path: path to file containing information on gene locations\n",
    "    ===\n",
    "    return a dataframe with genes as rows and samples as columns\n",
    "    make sure the genes are contained in the chromosome location info file (take intersection)\n",
    "    \"\"\"\n",
    "    # read in data\n",
    "    sc_df = pd.read_csv(data_path,index_col=0)\n",
    "    sc_df = sc_df.loc[[idx for idx in sc_df.index if 'ENSMUSG' in idx]]\n",
    "    gloc = pd.read_csv(gene_loc_path,sep=\"\\t\",index_col=0)\n",
    "    tad = pd.read_csv(tad_path)\n",
    "    # index manipulation\n",
    "    sc_df_idx = [idx.split(\".\")[0] for idx in sc_df.index]\n",
    "    sc_df.index = sc_df_idx\n",
    "    gloc_idx = [idx.split(\".\")[0] for idx in gloc.index]\n",
    "    gloc.index = gloc_idx\n",
    "    # get rid of genes with 0 exp across all samples\n",
    "    sc_df_filtered = sc_df.loc[np.sum(sc_df,axis=1)!=0]\n",
    "    # get rid of chromosomes in exclusion list\n",
    "    gloc_filtered = gloc[gloc['seqname'].isin(set(gloc.seqname).difference(excl_chrom))]\n",
    "    tad_filtered = tad[tad['chrom'].isin(set(gloc.seqname).difference(excl_chrom))]\n",
    "    # get intersecting genes in both\n",
    "    gene_list = set(sc_df_filtered.index).intersection(gloc_filtered.index)\n",
    "    sc_df_filtered = sc_df_filtered.loc[gene_list]\n",
    "    gloc_filtered = gloc_filtered.loc[gene_list]\n",
    "    return sc_df_filtered, gloc_filtered, tad_filtered\n",
    "\n",
    "def chromosome_gene_dict(gloc_data):\n",
    "    \"\"\"\n",
    "    return dictionary where keys are chromosomes and values are gene lists\n",
    "    \"\"\"\n",
    "    chromosomes = set(gloc_data.seqname)\n",
    "    cg_dict = dict(zip(chromosomes, [None]*len(chromosomes)))\n",
    "    for chrom in chromosomes:\n",
    "        gloc = gloc_data[gloc_data.seqname==chrom]\n",
    "        cg_dict[chrom] = list(gloc.index)\n",
    "    return cg_dict\n",
    "\n",
    "def get_genes_in_interval(chrom,start,end,gloc):\n",
    "    \"\"\"\n",
    "    get genes in an interval on a chromosome\n",
    "    \"\"\"\n",
    "    gloc_chr = gloc[gloc['seqname']==chrom]\n",
    "    gloc_chr = gloc_chr[(gloc_chr['start'] >= start) & (gloc_chr['end'] < end)]\n",
    "    return list(gloc_chr.index)\n",
    "\n",
    "def tad_gene_dict(tad_locs,gloc,filter_bar=5):\n",
    "    tg_dict = dict(zip(list(range(len(tad_locs))),[[]]*len(tad_locs)))\n",
    "    for i in range(len(tad_locs)):\n",
    "        data = tad_locs.loc[i]\n",
    "        tg_dict[i] = get_genes_in_interval(data['chrom'],data['start'],data['end'],gloc)\n",
    "    if filter_bar > 0:\n",
    "        tg_dict = {k:v for k,v in tg_dict.items() if len(v)>=filter_bar}\n",
    "    return tg_dict\n",
    "\n",
    "def log2norm_tpm(tpm_data):\n",
    "    \"\"\"\n",
    "    returns log2-normalized tpm data\n",
    "    \"\"\"\n",
    "    return zscore(np.log2(tpm_data+1),axis=1)\n",
    "\n",
    "def get_genes_from_chromosome(chr_name,tpm_data,tad_data,gloc_data):\n",
    "    \"\"\"\n",
    "    chr_name: e.g. 'chr1', the string corresponding to the chromosome you want to extract data on\n",
    "    tpm_data: dataframe of tpms\n",
    "    gene_loc_data: dataframe of gene locations\n",
    "    ===\n",
    "    returns filtered dataframes corresponding to chromosome of interest\n",
    "    \"\"\"\n",
    "    gloc_filtered = gloc_data[gloc_data['seqname']==chr_name]\n",
    "    tad_filtered = tad_data[tad_data['chrom']==chr_name]\n",
    "    genes = gloc_filtered.index\n",
    "    return tpm_data.loc[genes], gloc_filtered, tad_filtered\n",
    "\n",
    "def get_chr_lengths(path_to_file=\"../data/chr_lengths\"):\n",
    "    \"\"\"\n",
    "    return a dict of chromosomes and their lengths\n",
    "    \"\"\"\n",
    "    chr_lengths = {}\n",
    "    infile = open(path_to_file)\n",
    "    for line in infile:\n",
    "        line = line.strip().split()\n",
    "        chr = line[0]\n",
    "        length = int(line[1])\n",
    "        chr_lengths[chr] = length\n",
    "    infile.close()\n",
    "    return chr_lengths\n",
    "\n",
    "def slide_boundary(chr, start, end, num_iter=5, x=0.2):\n",
    "    \"\"\"\n",
    "    shift a TAD boundary left or right [num_iter] times with a step size of [x] * TAD size \n",
    "    while retaining the size of the TAD. return a list of new boundaries and the\n",
    "    corresponding distances from the original TAD.\n",
    "    \"\"\"\n",
    "    chr_length = get_chr_lengths()[chr]\n",
    "    # set step size\n",
    "    tad_length = end - start\n",
    "    step_size = 10000 # tad_length * x\n",
    "    # default: shift right\n",
    "    direction = 1\n",
    "    # TAD at end: shift left\n",
    "    if end + (step_size * num_iter) > chr_length:\n",
    "        direction = -1\n",
    "    new_boundaries = []\n",
    "    distances = []\n",
    "    step_size = direction * step_size\n",
    "    distance_from_origin = step_size\n",
    "    for i in range(num_iter):\n",
    "        new_start, new_end = start + step_size, end + step_size\n",
    "        new_boundaries.append((chr, new_start, new_end))\n",
    "        distances.append(distance_from_origin)\n",
    "        start, end = new_start, new_end\n",
    "        distance_from_origin += step_size\n",
    "    return new_boundaries, distances\n",
    "\n",
    "def calc_tad_coexp(chr, start, end, gene_loc, tpm):\n",
    "    \"\"\"\n",
    "    calculate the average pairwise correlation coefficient for a given genomic region\n",
    "    \"\"\"\n",
    "    genes = get_genes_in_interval(chr, start, end, gene_loc)\n",
    "    if len(genes) == 0:\n",
    "        return None, 0\n",
    "    tpm_subset = tpm.loc[genes,:]\n",
    "    corr_df = tpm_subset.transpose().corr()\n",
    "    avg_corr = np.mean(corr_df.to_numpy())\n",
    "    return corr_df, avg_corr\n",
    "\n",
    "def plot_corr_distance(title, distances, correlation, is_tad=True):\n",
    "    \"\"\"\n",
    "    plot correlation as a function of distance (either distance from TAD boundary or distance between gene pairs)\n",
    "    \"\"\"\n",
    "    plt.plot(distances, correlation)\n",
    "    if is_tad:\n",
    "        plt.title(\"Average pairwise correlation\")\n",
    "        plt.xlabel(\"Distance from TAD boundary in kb\")\n",
    "        plt.ylabel(\"Average correlation\")\n",
    "    else:\n",
    "        plt.title(\"Pairwise correlation\")\n",
    "        plt.xlabel(\"Distance between genes in bp\")\n",
    "        plt.ylabel(\"Correlation\")\n",
    "    # plt.savefig(title + \"_lineplot.png\")\n",
    "    plt.show()\n",
    "\n",
    "def plot_tad_heatmap(title, corr_df):\n",
    "    \"\"\"\n",
    "    plot heatmap of gene correlation within a TAD\n",
    "    \"\"\"\n",
    "    sns.clustermap(corr_df)\n",
    "    plt.show()\n",
    "\n",
    "def get_highly_correlated_genes(corr_df, percentile=99):\n",
    "    \"\"\"\n",
    "    given a correlation dataframe, return a list of the most highly correlated gene pairs.\n",
    "    \"\"\"\n",
    "    corr_matrix = corr_df.to_numpy()\n",
    "    values = corr_matrix[np.triu_indices_from(corr_matrix, 1)].flatten()\n",
    "    threshold = np.percentile(values, percentile)\n",
    "    genes = corr_df.index\n",
    "    all_pairs = list(itertools.combinations(genes, 2))\n",
    "    highly_correlated_pairs = []\n",
    "    for pair in all_pairs:\n",
    "        gene1, gene2 = pair[0], pair[1]\n",
    "        corr = corr_df.loc[gene1,gene2]\n",
    "        if corr > threshold:\n",
    "            highly_correlated_pairs.append((gene1, gene2))\n",
    "    return highly_correlated_pairs\n",
    "\n",
    "def calc_gene_dist(same_chrom_gene_pair,gene_loc):\n",
    "    \"\"\"\n",
    "    Calculate gene distance assuming genes are on the same chromosome\n",
    "    Midpoint distance\n",
    "    \"\"\"\n",
    "    gene1 = same_chrom_gene_pair[0]\n",
    "    gene2 = same_chrom_gene_pair[1]\n",
    "    gene1_srt = gene_loc.loc[gene1]['start']\n",
    "    gene1_end = gene_loc.loc[gene1]['end']\n",
    "    gene2_srt = gene_loc.loc[gene2]['start']\n",
    "    gene2_end = gene_loc.loc[gene2]['end']\n",
    "    return abs((gene1_srt+gene1_end)/2-(gene2_srt+gene2_end)/2)\n",
    "\n",
    "def genes_in_same_tad(gene_pair,tg_dict):\n",
    "    \"\"\"\n",
    "    given a pair of genes as a tuple, determine whether the genes are in the same TAD.\n",
    "    \"\"\"\n",
    "    for v in tg_dict.values():\n",
    "        if gene_pair[0] in v:\n",
    "            if gene_pair[1] in v:\n",
    "                return True\n",
    "            else:\n",
    "                return False\n",
    "    return False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/gn/1dwlg7ns6s3d_22hhl8kmrwm0000gn/T/ipykernel_3183/913813786.py:41: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  sc_df_filtered = sc_df_filtered.loc[gene_list]\n",
      "/var/folders/gn/1dwlg7ns6s3d_22hhl8kmrwm0000gn/T/ipykernel_3183/913813786.py:42: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  gloc_filtered = gloc_filtered.loc[gene_list]\n"
     ]
    }
   ],
   "source": [
    "# read in data\n",
    "tpm_data, gene_loc_data, tad_data = extract_data(data_path, gene_loc_path, tad_path)\n",
    "# normalize tpm data\n",
    "norm_tpm = log2norm_tpm(tpm_data)\n",
    "# make dictionaries\n",
    "cg_dict = chromosome_gene_dict(gene_loc_data)\n",
    "tg_dict = tad_gene_dict(tad_data,gene_loc_data)\n",
    "\n",
    "chromosome_list = ['chr'+str(i+1) for i in range(19)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for chromosome in chromosome_list:\n",
    "#     tpm, gene_loc, tad = get_genes_from_chromosome(chromosome,norm_tpm,tad_data,gene_loc_data)\n",
    "#     for t in tad.index:\n",
    "#         if t not in tg_dict.keys():\n",
    "#             continue\n",
    "#         tad_chr, tad_start, tad_end = tad.loc[t,:]\n",
    "#         tad_corr_df, tad_corr = calc_tad_coexp(tad_chr, tad_start, tad_end, gene_loc_data, tpm)\n",
    "#         new_boundaries, slide_distances = slide_boundary(tad_chr, tad_start, tad_end, 20)\n",
    "#         corr = [tad_corr]\n",
    "#         distances = [0] + slide_distances\n",
    "#         for new_chr, new_start, new_end in new_boundaries:\n",
    "#             new_corr_df, new_corr = calc_tad_coexp(new_chr, new_start, new_end, gene_loc_data, tpm)\n",
    "#             corr.append(new_corr)\n",
    "#         tad_location = \"{}_{}-{}\".format(tad_chr, tad_start, tad_end)\n",
    "#         plot_tad_heatmap(tad_location, tad_corr_df)\n",
    "#         plot_corr_distance(tad_location, np.array(distances)/1000 , corr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "def high_low_corr_genes(corr_df, percentile, high=True):\n",
    "    \"\"\"\n",
    "    if high, return all gene pairs greater than percentile\n",
    "    if low, return all gene pairs less than percentile\n",
    "    \"\"\"\n",
    "    corr_matrix = corr_df.to_numpy()\n",
    "    values = corr_matrix[np.triu_indices_from(corr_matrix, 1)].flatten()\n",
    "    threshold = np.percentile(values, percentile)\n",
    "    genes = corr_df.index\n",
    "    all_pairs = list(itertools.combinations(genes, 2))\n",
    "    correlated_gene_pairs = []\n",
    "    for pair in all_pairs:\n",
    "        gene1, gene2 = pair[0], pair[1]\n",
    "        corr = corr_df.loc[gene1,gene2]\n",
    "        if high:\n",
    "            if corr > threshold and corr < 0.99:\n",
    "                correlated_gene_pairs.append((gene1,gene2))\n",
    "        else:\n",
    "            if corr < threshold:\n",
    "                correlated_gene_pairs.append((gene1,gene2))\n",
    "    return correlated_gene_pairs\n",
    "    # pairs_in_tad = []\n",
    "    # print(len(correlated_gene_pairs))\n",
    "    # for pair in correlated_gene_pairs:\n",
    "    #     if genes_in_same_tad(pair, tg_dict):\n",
    "    #         pairs_in_tad.append(pair)\n",
    "    # return len(pairs_in_tad) / len(correlated_gene_pairs) * 100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ENSMUSG00000089493</th>\n",
       "      <th>ENSMUSG00000027712</th>\n",
       "      <th>ENSMUSG00000102218</th>\n",
       "      <th>ENSMUSG00000040016</th>\n",
       "      <th>ENSMUSG00000049796</th>\n",
       "      <th>ENSMUSG00000028145</th>\n",
       "      <th>ENSMUSG00000040600</th>\n",
       "      <th>ENSMUSG00000102846</th>\n",
       "      <th>ENSMUSG00000070372</th>\n",
       "      <th>ENSMUSG00000028081</th>\n",
       "      <th>...</th>\n",
       "      <th>ENSMUSG00000053769</th>\n",
       "      <th>ENSMUSG00000104822</th>\n",
       "      <th>ENSMUSG00000105449</th>\n",
       "      <th>ENSMUSG00000027997</th>\n",
       "      <th>ENSMUSG00000037211</th>\n",
       "      <th>ENSMUSG00000049128</th>\n",
       "      <th>ENSMUSG00000097583</th>\n",
       "      <th>ENSMUSG00000091405</th>\n",
       "      <th>ENSMUSG00000106462</th>\n",
       "      <th>ENSMUSG00000105279</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000089493</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>-0.607134</td>\n",
       "      <td>0.401431</td>\n",
       "      <td>-0.430653</td>\n",
       "      <td>0.294673</td>\n",
       "      <td>0.474112</td>\n",
       "      <td>-0.199498</td>\n",
       "      <td>0.337268</td>\n",
       "      <td>-0.015633</td>\n",
       "      <td>-0.422481</td>\n",
       "      <td>...</td>\n",
       "      <td>0.550158</td>\n",
       "      <td>0.279538</td>\n",
       "      <td>0.165525</td>\n",
       "      <td>-0.388498</td>\n",
       "      <td>-0.370888</td>\n",
       "      <td>-0.248973</td>\n",
       "      <td>0.538285</td>\n",
       "      <td>-0.238905</td>\n",
       "      <td>0.617816</td>\n",
       "      <td>-0.134472</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000027712</th>\n",
       "      <td>-0.607134</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-0.179539</td>\n",
       "      <td>0.675815</td>\n",
       "      <td>-0.148466</td>\n",
       "      <td>-0.584874</td>\n",
       "      <td>0.178609</td>\n",
       "      <td>-0.478927</td>\n",
       "      <td>0.557784</td>\n",
       "      <td>0.818824</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.615915</td>\n",
       "      <td>-0.093476</td>\n",
       "      <td>-0.127600</td>\n",
       "      <td>0.760592</td>\n",
       "      <td>0.740099</td>\n",
       "      <td>0.590021</td>\n",
       "      <td>-0.509723</td>\n",
       "      <td>0.136095</td>\n",
       "      <td>-0.371516</td>\n",
       "      <td>-0.300660</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000102218</th>\n",
       "      <td>0.401431</td>\n",
       "      <td>-0.179539</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-0.338086</td>\n",
       "      <td>-0.176866</td>\n",
       "      <td>0.047350</td>\n",
       "      <td>-0.196988</td>\n",
       "      <td>0.267740</td>\n",
       "      <td>0.152257</td>\n",
       "      <td>0.068454</td>\n",
       "      <td>...</td>\n",
       "      <td>0.158865</td>\n",
       "      <td>0.192704</td>\n",
       "      <td>0.242292</td>\n",
       "      <td>0.025125</td>\n",
       "      <td>0.278828</td>\n",
       "      <td>-0.152617</td>\n",
       "      <td>0.303522</td>\n",
       "      <td>0.068005</td>\n",
       "      <td>-0.134568</td>\n",
       "      <td>-0.134568</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000040016</th>\n",
       "      <td>-0.430653</td>\n",
       "      <td>0.675815</td>\n",
       "      <td>-0.338086</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.018216</td>\n",
       "      <td>-0.201007</td>\n",
       "      <td>0.558651</td>\n",
       "      <td>-0.258342</td>\n",
       "      <td>0.592005</td>\n",
       "      <td>0.703546</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.292282</td>\n",
       "      <td>0.093897</td>\n",
       "      <td>0.175123</td>\n",
       "      <td>0.669700</td>\n",
       "      <td>0.643580</td>\n",
       "      <td>0.404859</td>\n",
       "      <td>-0.267811</td>\n",
       "      <td>0.179591</td>\n",
       "      <td>-0.093638</td>\n",
       "      <td>-0.380587</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000049796</th>\n",
       "      <td>0.294673</td>\n",
       "      <td>-0.148466</td>\n",
       "      <td>-0.176866</td>\n",
       "      <td>0.018216</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.409292</td>\n",
       "      <td>-0.347019</td>\n",
       "      <td>-0.089880</td>\n",
       "      <td>0.231980</td>\n",
       "      <td>-0.323303</td>\n",
       "      <td>...</td>\n",
       "      <td>0.434670</td>\n",
       "      <td>-0.163697</td>\n",
       "      <td>0.161233</td>\n",
       "      <td>-0.273581</td>\n",
       "      <td>-0.164675</td>\n",
       "      <td>-0.383144</td>\n",
       "      <td>0.094180</td>\n",
       "      <td>-0.249552</td>\n",
       "      <td>0.433385</td>\n",
       "      <td>-0.238792</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 2133 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    ENSMUSG00000089493  ENSMUSG00000027712  \\\n",
       "ENSMUSG00000089493            1.000000           -0.607134   \n",
       "ENSMUSG00000027712           -0.607134            1.000000   \n",
       "ENSMUSG00000102218            0.401431           -0.179539   \n",
       "ENSMUSG00000040016           -0.430653            0.675815   \n",
       "ENSMUSG00000049796            0.294673           -0.148466   \n",
       "\n",
       "                    ENSMUSG00000102218  ENSMUSG00000040016  \\\n",
       "ENSMUSG00000089493            0.401431           -0.430653   \n",
       "ENSMUSG00000027712           -0.179539            0.675815   \n",
       "ENSMUSG00000102218            1.000000           -0.338086   \n",
       "ENSMUSG00000040016           -0.338086            1.000000   \n",
       "ENSMUSG00000049796           -0.176866            0.018216   \n",
       "\n",
       "                    ENSMUSG00000049796  ENSMUSG00000028145  \\\n",
       "ENSMUSG00000089493            0.294673            0.474112   \n",
       "ENSMUSG00000027712           -0.148466           -0.584874   \n",
       "ENSMUSG00000102218           -0.176866            0.047350   \n",
       "ENSMUSG00000040016            0.018216           -0.201007   \n",
       "ENSMUSG00000049796            1.000000            0.409292   \n",
       "\n",
       "                    ENSMUSG00000040600  ENSMUSG00000102846  \\\n",
       "ENSMUSG00000089493           -0.199498            0.337268   \n",
       "ENSMUSG00000027712            0.178609           -0.478927   \n",
       "ENSMUSG00000102218           -0.196988            0.267740   \n",
       "ENSMUSG00000040016            0.558651           -0.258342   \n",
       "ENSMUSG00000049796           -0.347019           -0.089880   \n",
       "\n",
       "                    ENSMUSG00000070372  ENSMUSG00000028081  ...  \\\n",
       "ENSMUSG00000089493           -0.015633           -0.422481  ...   \n",
       "ENSMUSG00000027712            0.557784            0.818824  ...   \n",
       "ENSMUSG00000102218            0.152257            0.068454  ...   \n",
       "ENSMUSG00000040016            0.592005            0.703546  ...   \n",
       "ENSMUSG00000049796            0.231980           -0.323303  ...   \n",
       "\n",
       "                    ENSMUSG00000053769  ENSMUSG00000104822  \\\n",
       "ENSMUSG00000089493            0.550158            0.279538   \n",
       "ENSMUSG00000027712           -0.615915           -0.093476   \n",
       "ENSMUSG00000102218            0.158865            0.192704   \n",
       "ENSMUSG00000040016           -0.292282            0.093897   \n",
       "ENSMUSG00000049796            0.434670           -0.163697   \n",
       "\n",
       "                    ENSMUSG00000105449  ENSMUSG00000027997  \\\n",
       "ENSMUSG00000089493            0.165525           -0.388498   \n",
       "ENSMUSG00000027712           -0.127600            0.760592   \n",
       "ENSMUSG00000102218            0.242292            0.025125   \n",
       "ENSMUSG00000040016            0.175123            0.669700   \n",
       "ENSMUSG00000049796            0.161233           -0.273581   \n",
       "\n",
       "                    ENSMUSG00000037211  ENSMUSG00000049128  \\\n",
       "ENSMUSG00000089493           -0.370888           -0.248973   \n",
       "ENSMUSG00000027712            0.740099            0.590021   \n",
       "ENSMUSG00000102218            0.278828           -0.152617   \n",
       "ENSMUSG00000040016            0.643580            0.404859   \n",
       "ENSMUSG00000049796           -0.164675           -0.383144   \n",
       "\n",
       "                    ENSMUSG00000097583  ENSMUSG00000091405  \\\n",
       "ENSMUSG00000089493            0.538285           -0.238905   \n",
       "ENSMUSG00000027712           -0.509723            0.136095   \n",
       "ENSMUSG00000102218            0.303522            0.068005   \n",
       "ENSMUSG00000040016           -0.267811            0.179591   \n",
       "ENSMUSG00000049796            0.094180           -0.249552   \n",
       "\n",
       "                    ENSMUSG00000106462  ENSMUSG00000105279  \n",
       "ENSMUSG00000089493            0.617816           -0.134472  \n",
       "ENSMUSG00000027712           -0.371516           -0.300660  \n",
       "ENSMUSG00000102218           -0.134568           -0.134568  \n",
       "ENSMUSG00000040016           -0.093638           -0.380587  \n",
       "ENSMUSG00000049796            0.433385           -0.238792  \n",
       "\n",
       "[5 rows x 2133 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chromosome = \"chr3\"\n",
    "tpm, gene_loc, tad = get_genes_from_chromosome(chromosome,norm_tpm,tad_data,gene_loc_data)\n",
    "all_genes_corr_df = tpm.transpose().corr()\n",
    "all_genes_corr_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "22738\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0.9411557744744481"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# top 99 percentile\n",
    "high_low_corr_genes(all_genes_corr_df, 99, True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "22738\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0.07916263523616854"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# bottom 1 percentile\n",
    "high_low_corr_genes(all_genes_corr_df, 1, False)"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "243639d3113cbe381352a32fe5a4e1a317b20dd77d7d45bde22354ae8ddc7d11"
  },
  "kernelspec": {
   "display_name": "Python 3.9.7 ('base')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
