{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5cfd47a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from scipy.stats import zscore\n",
    "import argparse\n",
    "import itertools\n",
    "from tqdm import tqdm\n",
    "import random\n",
    "import warnings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "18d8d8e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_data(data_path,gene_loc_path,tad_path,hivar_pctl=None,excl_chrom=['chrM','chrX','chrY']):\n",
    "    \"\"\"\n",
    "    excl_chrom: list of strings corresponding to chromosomes to be excluded\n",
    "    data_path: path to single cell tpm data\n",
    "    gene_loc_path: path to file containing information on gene locations\n",
    "    ===\n",
    "    return a dataframe with genes as rows and samples as columns\n",
    "    make sure the genes are contained in the chromosome location info file (take intersection)\n",
    "    \"\"\"\n",
    "    # read in data\n",
    "    sc_df = pd.read_csv(data_path,index_col=0)\n",
    "    sc_df = sc_df.loc[[idx for idx in sc_df.index if 'ENSMUSG' in idx]]\n",
    "    gloc = pd.read_csv(gene_loc_path,sep=\"\\t\",index_col=0)\n",
    "    tad = pd.read_csv(tad_path)\n",
    "    # index manipulation\n",
    "    sc_df_idx = [idx.split(\".\")[0] for idx in sc_df.index]\n",
    "    sc_df.index = sc_df_idx\n",
    "    gloc_idx = [idx.split(\".\")[0] for idx in gloc.index]\n",
    "    gloc.index = gloc_idx\n",
    "    # get rid of genes with 0 exp across all samples\n",
    "    sc_df_filtered = sc_df.loc[np.sum(sc_df,axis=1)!=0]\n",
    "    # filter gene by variance across samples/tissues if thus specified\n",
    "    if type(hivar_pctl)==int or type(hivar_pctl)==float:\n",
    "        if hivar_pctl>=0 and hivar_pctl<=100:\n",
    "            sc_df_filtered = filter_genes_by_variance(sc_df_filtered,hivar_pctl)\n",
    "    # get rid of chromosomes in exclusion list\n",
    "    gloc_filtered = gloc[gloc['seqname'].isin(set(gloc.seqname).difference(excl_chrom))]\n",
    "    tad_filtered = tad[tad['chrom'].isin(set(gloc.seqname).difference(excl_chrom))]\n",
    "    # get intersecting genes in both\n",
    "    gene_list = set(sc_df_filtered.index).intersection(gloc_filtered.index)\n",
    "    sc_df_filtered = sc_df_filtered.loc[gene_list]\n",
    "    gloc_filtered = gloc_filtered.loc[gene_list]\n",
    "    return sc_df_filtered, gloc_filtered, tad_filtered\n",
    "def filter_genes_by_variance(sc_df,percentile):\n",
    "    \"\"\"\n",
    "    filters a TPM (genes x tissue !!!!) dataset by highest percentile of variance across tissue/samples\n",
    "\n",
    "    IMPORTANT: do not feed in normalized data. That would make this pointless.\n",
    "    \"\"\"\n",
    "    sc_df = sc_df.T\n",
    "    gene_vars = sc_df.var()\n",
    "    most_var_genes = (gene_vars >= np.percentile(gene_vars,percentile))\n",
    "    return sc_df.loc[:,most_var_genes].T # inverts back to genes x samples/tissues\n",
    "\n",
    "\n",
    "def log2norm_tpm(tpm_data):\n",
    "    \"\"\"\n",
    "    returns log2-normalized tpm data\n",
    "    \"\"\"\n",
    "    return zscore(np.log2(tpm_data+1),axis=1)\n",
    "\n",
    "\n",
    "def tad_gene_dict(tad_locs,gloc,filter_bar=5):\n",
    "    tg_dict = dict(zip(list(range(len(tad_locs))),[[]]*len(tad_locs)))\n",
    "    for i in range(len(tad_locs)):\n",
    "        data = tad_locs.loc[i]\n",
    "        tg_dict[i] = get_genes_in_interval(data['chrom'],data['start'],data['end'],gloc)\n",
    "    if filter_bar > 0:\n",
    "        tg_dict = {k:v for k,v in tg_dict.items() if len(v)>=filter_bar}\n",
    "    return tg_dict\n",
    "def get_genes_in_interval(chrom,start,end,gloc):\n",
    "    \"\"\"\n",
    "    get genes in an interval on a chromosome\n",
    "    \"\"\"\n",
    "    gloc_chr = gloc[gloc['seqname']==chrom]\n",
    "    gloc_chr = gloc_chr[(gloc_chr['start'] >= start) & (gloc_chr['end'] < end)]\n",
    "    return list(gloc_chr.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a74f9050",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_genes_from_chromosome(chr_name,tpm_data,tad_data,gloc_data):\n",
    "    \"\"\"\n",
    "    chr_name: e.g. 'chr1', the string corresponding to the chromosome you want to extract data on\n",
    "    tpm_data: dataframe of tpms\n",
    "    gene_loc_data: dataframe of gene locations\n",
    "    ===\n",
    "    returns filtered dataframes corresponding to chromosome of interest\n",
    "    \"\"\"\n",
    "    gloc_filtered = gloc_data[gloc_data['seqname']==chr_name]\n",
    "    tad_filtered = tad_data[tad_data['chrom']==chr_name]\n",
    "    genes = gloc_filtered.index\n",
    "    return tpm_data.loc[genes], gloc_filtered, tad_filtered\n",
    "\n",
    "def chromosome_gene_dict(gloc_data):\n",
    "    \"\"\"\n",
    "    return dictionary where keys are chromosomes and values are gene lists\n",
    "    \"\"\"\n",
    "    chromosomes = set(gloc_data.seqname)\n",
    "    cg_dict = dict(zip(chromosomes, [None]*len(chromosomes)))\n",
    "    for chrom in chromosomes:\n",
    "        gloc = gloc_data[gloc_data.seqname==chrom]\n",
    "        cg_dict[chrom] = list(gloc.index)\n",
    "    return cg_dict\n",
    "\n",
    "def get_chr_lengths(path_to_file=\"../data/chr_lengths\"):\n",
    "    \"\"\"\n",
    "    return a dict of chromosomes and their lengths\n",
    "    \"\"\"\n",
    "    chr_lengths = {}\n",
    "    infile = open(path_to_file)\n",
    "    for line in infile:\n",
    "        line = line.strip().split()\n",
    "        chrom = line[0]\n",
    "        length = int(line[1])\n",
    "        chr_lengths[chrom] = length\n",
    "    infile.close()\n",
    "    return chr_lengths\n",
    "\n",
    "def gene_corr_dictionary(tpm):\n",
    "    gcorr_dict = {}\n",
    "    corr_matrix = tpm.T.corr()\n",
    "    for gene_pair in itertools.combinations(tpm.index,2):\n",
    "        gcorr_dict[gene_pair] = corr_matrix[gene_pair[0]][gene_pair[1]]\n",
    "    return gcorr_dict\n",
    "\n",
    "def genes_in_same_tad(gene_pair,tg_dict,return_false_if_same_genes=True):\n",
    "    \"\"\"\n",
    "    given a pair of genes as a tuple, determine whether the genes are in the same TAD.\n",
    "    \"\"\"\n",
    "    if return_false_if_same_genes: # whether to treat same gene as belonging to the same TAD\n",
    "        if gene_pair[0]==gene_pair[1]:\n",
    "            return False\n",
    "    for v in tg_dict.values():\n",
    "        if gene_pair[0] in v:\n",
    "            if gene_pair[1] in v:\n",
    "                return True\n",
    "            else:\n",
    "                return False\n",
    "    return False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "72b12cee",
   "metadata": {},
   "outputs": [],
   "source": [
    "def gene_tad_dict(gene_pairs, tg_dict):\n",
    "    \"\"\"\n",
    "    Returns dictionary indicating whether a gene pair shares a TAD\n",
    "    e.g. {(gene1, gene2) : True/False}\n",
    "    \"\"\"\n",
    "    gt_dict = {}\n",
    "    for pair in gene_pairs:\n",
    "        gt_dict[pair] = genes_in_same_tad(pair,tg_dict)\n",
    "    return gt_dict\n",
    "def gene_corr_permutation_test(gc_dict, tg_dict, percentile=90, n_iter=100, rand_seed=1):\n",
    "    \"\"\"\n",
    "    Given a dictionary of gene pairs mapping to correlations, get top 90th percentile\n",
    "    gc_dict should not have autocorrelation key-value pairs\n",
    "    \"\"\"\n",
    "    np.random.seed(rand_seed)\n",
    "    \n",
    "    correlations = list(gc_dict.values())\n",
    "    gpairs = list(gc_dict.keys())\n",
    "    threshold = np.percentile(correlations, percentile)\n",
    "    hi_corr_pairs = [gp for gp in gpairs if gc_dict[gp]>threshold]\n",
    "    gt_dict = gene_tad_dict(gpairs, tg_dict)\n",
    "    \n",
    "    if sum(gt_dict.values()) == 0:\n",
    "        return None # everything will just be 0; no point\n",
    "    \n",
    "    proportion = sum([gt_dict[p] for p in hi_corr_pairs])/len(hi_corr_pairs)\n",
    "    \n",
    "    sample_size = len(hi_corr_pairs)\n",
    "    proportions = np.ones(n_iter) # proportion in TAD\n",
    "    \n",
    "    for i in range(n_iter):\n",
    "        sample = random.sample(gpairs,sample_size)\n",
    "        proportions[i] = sum(gt_dict[gpair] for gpair in sample)/sample_size\n",
    "    \n",
    "    return sum(proportions>proportion)/n_iter # p value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "6adb676d",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path = \"../data/ENCODE_bulk_rna_seq.csv\"\n",
    "gene_loc_path = \"../data/gene_locations.tsv\"\n",
    "tad_path = \"../data/TAD_strong_boundary_start_end.csv\"\n",
    "\n",
    "min_genes_in_tad = 5 #30\n",
    "\n",
    "\n",
    "# read in data\n",
    "tpm_data, gene_loc_data, tad_data = extract_data(data_path, gene_loc_path, tad_path)\n",
    "# normalize tpm data\n",
    "norm_tpm = log2norm_tpm(tpm_data)\n",
    "# make dictionaries\n",
    "#cg_dict = chromosome_gene_dict(gene_loc_data)\n",
    "tg_dict = tad_gene_dict(tad_data,gene_loc_data,filter_bar=min_genes_in_tad)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c94ae2ea",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing chr1...\n",
      "Analysis 4 done.\n",
      "Processing chr2...\n",
      "Analysis 4 done.\n",
      "Processing chr3...\n",
      "Analysis 4 done.\n",
      "Processing chr4...\n",
      "Analysis 4 done.\n",
      "Processing chr5...\n",
      "Analysis 4 done.\n",
      "Processing chr6...\n",
      "Analysis 4 done.\n",
      "Processing chr7...\n",
      "Analysis 4 done.\n",
      "Processing chr8...\n",
      "Analysis 4 done.\n",
      "Processing chr9...\n",
      "Analysis 4 done.\n",
      "Processing chr10...\n",
      "Analysis 4 done.\n",
      "Processing chr11...\n",
      "Analysis 4 done.\n",
      "Processing chr12...\n",
      "Analysis 4 done.\n",
      "Processing chr13...\n",
      "Analysis 4 done.\n",
      "Processing chr14...\n",
      "Analysis 4 done.\n",
      "Processing chr15...\n",
      "Analysis 4 done.\n",
      "Processing chr16...\n",
      "Analysis 4 done.\n",
      "Processing chr17...\n",
      "Analysis 4 done.\n",
      "Processing chr18...\n",
      "Analysis 4 done.\n",
      "Processing chr19...\n"
     ]
    }
   ],
   "source": [
    "chromosome_list = ['chr'+str(i+1) for i in range(19)]\n",
    "\n",
    "for chromosome in chromosome_list:\n",
    "    print(f\"Processing {chromosome}...\")\n",
    "    tpm, gene_loc, tad = get_genes_from_chromosome(chromosome,norm_tpm,tad_data,gene_loc_data)\n",
    "    gc_dict = gene_corr_dictionary(tpm)\n",
    "    p_val = gene_corr_permutation_test(gc_dict, tg_dict)\n",
    "    with open(\"../results/pvalues.txt\",'a') as f:\n",
    "        f.write(chromosome+\"\\t\"+str(p_val)+\"\\n\")\n",
    "    f.close()\n",
    "    print(f\"Analysis 4 done.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bb9fc6cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "def high_low_corr_genes2(corr_df, margin=10):\n",
    "    \"\"\"\n",
    "    return all gene pairs greater than percentile\n",
    "    and all gene pairs less than percentile\n",
    "    and all genes in bottom (1-percentile) percentile\n",
    "    \"\"\"\n",
    "    corr_matrix = corr_df.to_numpy()\n",
    "    values = corr_matrix[np.triu_indices_from(corr_matrix, 1)].flatten()\n",
    "    threshold = np.percentile(values, 100-margin) # default 90th percentile\n",
    "    threshold_lower = np.percentile(values, margin) # default 10th percentile\n",
    "    all_pairs = get_gene_pairs(corr_df.index)\n",
    "    correlated_gene_pairs = []\n",
    "    low_corr_gene_pairs = []\n",
    "    neg_corr_gene_pairs = []\n",
    "    for pair in all_pairs:\n",
    "        gene1, gene2 = pair[0], pair[1]\n",
    "        corr = corr_df.loc[gene1,gene2]\n",
    "        if corr > threshold and corr < 1:\n",
    "            correlated_gene_pairs.append(pair)\n",
    "        elif corr < threshold_lower and corr > -1:\n",
    "            neg_corr_gene_pairs.append(pair)\n",
    "        elif corr <= threshold and corr >= threshold_lower:\n",
    "            low_corr_gene_pairs.append(pair)\n",
    "    return correlated_gene_pairs, low_corr_gene_pairs, neg_corr_gene_pairs\n",
    "def get_gene_pairs(genes):\n",
    "    \"\"\"\n",
    "    given a list of genes, return all possible pairs (not including self pairs)\n",
    "    \"\"\"\n",
    "    return list(itertools.combinations(genes, 2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bab44671",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ignore warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "for chromosome in chromosome_list:\n",
    "    print(f\"Processing {chromosome}...\")\n",
    "    tpm, gene_loc, tad = get_genes_from_chromosome(chromosome,norm_tpm,tad_data,gene_loc_data)\n",
    "    # analysis 1: correlation vs sliding windows\n",
    "    sliding_windows(tpm, gene_loc, tad, plot=True)\n",
    "    print(\"Analysis 1 done.\")\n",
    "    # analysis 2: are highly correlated genes in the same TAD?\n",
    "    all_genes_corr_df = tpm.transpose().corr()\n",
    "    analysis2_df = hc_genes_in_tads(chromosome, all_genes_corr_df, gene_loc, tad, tg_dict, plot=True)\n",
    "    analysis2_df.to_csv(f\"../results/analysis2_{chromosome}_df.csv\")\n",
    "    print(\"Analysis 2 done.\")\n",
    "    # analysis 3: correlation as a function of distance between genes\n",
    "    analysis3_high_corr_df, analysis3_other_corr_df = corr_vs_dist(all_genes_corr_df, gene_loc_data, plot=True, title = f\"Gene distance between high correlated and low correlated genes \\n in {chromosome}\", save=f\"../results/corr_vs_dist_plot_{chromosome}.png\")\n",
    "    analysis3_high_corr_df.to_csv(f\"../results/analysis3_{chromosome}_high_corr_df.csv\")\n",
    "    analysis3_other_corr_df.to_csv(f\"../results/analysis3_{chromosome}_other_corr_df.csv\")\n",
    "    print(f\"Analysis 3 done.\")\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b9c38b10",
   "metadata": {},
   "outputs": [],
   "source": [
    "# analysis 5: do NMF-identified gene modules tend to be in the same TAD?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3eccce3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# analysis 6: do PCA component elements tend to be in the same TAD?"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
